New feature: Save Find All menu content into a new Editor.

MKadaner opened the pull request:
## Summary
New feature added to the Find All menu:
- Key #F2# copies all found lines into a new editor and closes the menu.
- Key #Shift+F2# copies all lines matching the current filter into a new editor and closes the menu.

## Checklist
- [X] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [ ] I have discussed this with project maintainers: <!-- add a link to the corresponding issue / discussion / forum topic here --> <br/>
If not checked, I accept that this work might be rejected in favor of a different great big ineffable plan.<br/>

## Details
Thanks to @alabuzhev for the main implementation idea. Everything else was simple.

HamRusTal commented on far/FarEng.hlf.m4:
```diff
@@ -3434,6 +3434,12 @@ string is changed to that of the found string. For example:
  The following key combinations are available in this menu:
 
+ #F2#
+ Copy all found lines into a new editor and close this menu.
```
Unfortunately, it's not clear from this phrase what exactly will be copied to the new editor: found file paths (full), verbatim lines presented in the dialog (relative/truncated), or text lines from the found files' content.

HamRusTal commented on far/FarEng.hlf.m4:
```diff
@@ -3434,6 +3434,12 @@ string is changed to that of the found string. For example:
+ Copy all found lines into a new editor and close this menu.
+
+ #Shift+F2#
+ Copy all lines matching the current filter into a new editor and close this menu.
```
Unfortunately, it's not clear what `the current filter` is and how this action is different from **`F2`**.

HamRusTal commented:
I left a couple of irrelevant questions here earlier today and just deleted these.
Sorry for the noise and likely false notifications. 🙄

johnd0e commented:
> Key #F2# copies all found lines into a new editor and closes the menu.

So the new editor content is kind of grep results?

MKadaner commented:
@HamRusTal, no problems.

> So the new editor content is kind of grep results?

Yes, exactly. And this defined the choice of the accelerator key. When I conceived this feature, I thought about it as of redirecting grep output to a file, so `F2` felt quite natural. Then, I realized that copying the found lines to an Editor makes much more sense in all respects.

Few UX questions:

1. Is `F2` still a good choice? Can you think of a more natural accelerator?
2. Should `F2` save visible (filtered) lines, and `Shift+F2` all found lines or vice versa? Currently, `F2` means "all," and "`Shift+F2` "filtered." But maybe `F2` should mean "save what you see," and `Shift+F2` the alternative?
3. Should the new Editor be opened in Save-to-SaveAs mode, i.e., prompt for the file name on the first save?
4. Should the new Editor be considered "dirty," i.e., ask to save on `Esc`?

johnd0e commented:
> Is F2 still a good choice? Can you think of a more natural accelerator?

F4, AltF4, ShiftF4

3. Yes

4. No

MKadaner commented:
Thank you, @johnd0e. `F4`, of course! And `Alt+F4` by association with the (right) `Alt` used to activate filtering in the menu. So:

- `F4` copies all lines to the new editor.
- `Alt+F4` copies filtered lines to the new editor.
- On the first save, the new editor will ask for the filename (default is, for example, `Far.Filterd.map`).
- The new editor may be closed without any confirmation and without saving.

I'll update the PR accordingly.

alabuzhev commented on far/editor.cpp:
```diff
@@ -3970,6 +3999,41 @@ void Editor::DoSearchReplace(const SearchReplaceDisposition Disposition)
+	for (const auto I : ListBox.GetItems()
+		| filter([FilterFlags](const auto& Item) { return !(Item.Flags & FilterFlags); })
+		| transform([](const auto& Item) { return std::any_cast<FindCoord>(Item.ComplexUserData).Line; })
+		| chunk_by(std::ranges::equal_to{})
```
`chunk_by` [requires](https://en.cppreference.com/w/cpp/compiler_support.html) gcc 13 / clang 18 / vs 2022 17.3, which are higher than our [current requirements](https://github.com/FarGroup/FarManager/blob/002d6c5e5f8a5f57ecff05a361cdcf55bcf0fa5d/far/headers.hpp#L40-L46).

Do we really need it here? Having `int LastLine = -1;` outside the loop and `if (I == LastLine) continue; LastLine = I;` inside it should achieve the same effect.

P.S. The requirements can probably be updated, but preferably not in this PR.

MKadaner replied:
Sure, I can do a simple loop. `chunk_by` here is a bit questionable performane-wise. I did not measure, but I have some intuition here.

What it takes to update requirements besides changing versions in the snippet you pointed? What is the highest we can reasonable jump to? I think keeping up with the latest compilers and language feature is a really good thing, not just a fad.

alabuzhev replied:
> What it takes to update requirements besides changing versions in the snippet you pointed?

Make sure that our CI pipelines have those versions or newer.

> What is the highest we can reasonable jump to?

I think we discussed that already somewhere.
We do try to keep up with the latest, but also try to support 2-3 previous versions, released in the last 2-3 years or so.

MKadaner replied:
Thanks. I'll try to upgrade the compiler (separately of course). Lower priority for me. GH actions are defined in [`build.yml`](https://github.com/FarGroup/FarManager/blob/master/.github/workflows/build.yml), right? What about appveyor?

alabuzhev commented on far/editor.cpp:
```diff
@@ -3970,6 +3999,41 @@ void Editor::DoSearchReplace(const SearchReplaceDisposition Disposition)
+		| transform([](auto&& chunk) { return chunk.front(); }))
+	{
+		NewEditor->InsertString(GetStringByNumber(I)->GetString(), NewEditor->LastLine());
+		NewEditor->TextChanged(true);
```
Do we need to call this N times?
Maybe once after the loop?

MKadaner replied:
It is inexpensive, but yes, it should be out of the loop.

Besides, in the version I am working on, I do not do it at all. Since, as we discussed above, I want to avoid asking to save the file if the user have not touched it, I do not mark it dirty. There are more things I am doing to achieve the behavior I want. I'll explain it separately.

alabuzhev commented on far/farlang.templ.m4:
```diff
@@ -7681,6 +7681,21 @@ upd:"Occurrences: {0}, lines: {1}"
 upd:"Occurrences: {0}, lines: {1}"
 upd:"Occurrences: {0}, lines: {1}"
 
+MEditSearchAllFileNameAppend
```
Do we need this?
I doubt that anyone will ever want to change it, and changing it can do more harm than good anyway.

MKadaner replied:
Do you mean to hard code the string "Filtered" instead? I can do that. Yeah, I thought somebody may want to translate it to their language. Out of curiosity, what harm can come out of it?

I'd rather keep it but I am not strong on it. Please let me know one way or the other.

alabuzhev replied:
Someone could put unsupported characters there and then wonder why it doesn't work.
On the other hand, we do have MNewFileName there already, so I won't insist.

MKadaner replied:
Good point. However, the name is used only in the "Save As" dialog, so one will immediately see what one is about to do. 😀

I'll keep it for now.

alabuzhev commented on far/vmenu.hpp:
```diff
@@ -226,6 +226,7 @@ class VMenu final: public Modal
 	// SelectPos == -1 & non-empty Items - everything is filtered
 	bool HasVisible() const { return SelectPos > -1 && !Items.empty(); }
 	int GetShowItemCount() const { return static_cast<int>(Items.size() - ItemHiddenCount); }
+	const std::vector<MenuItemEx>& GetItems() const { return Items; };
```
Maybe return `span<MenuItemEx const>`?

MKadaner replied:
Will do.

HamRusTal commented:
> I propose to add a new flag, `FFILEEDIT_EPHEMERAL`. Any concerns?

If this way gets taken, shouldn't adding this flag be a separate PR preceding the “Save Find All …”?

MKadaner commented:
> > I propose to add a new flag, FFILEEDIT_EPHEMERAL. Any concerns?
>
> If this way gets taken, shouldn't adding this flag be a separate PR preceding the “Save Find All …”?

If the flag is added in a separate PR, in that PR there will be no way for reviewers to play with this flag. (I test / debug them together anyway.)

I suggest reviewing both the flag and the feature together. When everything is fine, I can split PR into two if you still think it a good idea.

MKadaner commented:
@alabuzhev, gentle ping. Last week three days out of five I badly wanted this feature; almost was at the point to clone the repo on my work computer and make a wildcat build.

alabuzhev commented on far/fileedit.cpp:
```diff
@@ -864,6 +868,7 @@ long long FileEditor::VMProcess(int OpCode, void* vParam, long long iParam)
 		MacroEditState |= m_editor->m_Flags.Check(Editor::FEDITOR_LOCKMODE)?              9_bit : 0;
 		MacroEditState |= m_editor->EdOpt.PersistentBlocks?                               10_bit : 0;
 		MacroEditState |= !GetCanLoseFocus()?                                             11_bit : 0;
+		MacroEditState |= m_Flags.Check(FFILEEDIT_EPHEMERAL)?                             12_bit : 0;
```
Please do then, we can always add it later if requested.

MKadaner commented:
I am addressing a couple of reasonable SonarQube warnings. Please let me know if you want me to refactor the places where we have "more than 3 nested if|for|do|while|switch statements."

MKadaner commented:
I removed exposing `FFILEEDIT_EPHEMERAL` and addressed a couple of SonarQube warning.

As for the too deep nesting warnings, I did increase nesting in two places, both of which already had more than three levels. I experimented with some meaningful refactoring, but it did not go well: too much global state.

In both cases, I can move the body of `if (!m_Flags.Check(FFILEEDIT_EPHEMERAL))` conditional to new functions, thus technically not increasing cyclomatic complexity, but I think the result would become even worse. So, I decided to leave these places as is.

Please let me know if you want me to introduce meaningless functions. Otherwise, I'll squash the commits and call it a day.

 I can (try to) refactor `FileEditor::Init` in a separate PR if you want me to. I hope the priority would not be too high. 😀

alabuzhev commented:
Thank you

MKadaner commented:
Thank you!

nightroman commented:
Hello All, thank you for the promising feature.

A flaw? As far as I understand this pattern is used for new file names with results: `{original-base-name}.Filtered.{original-extension}`. With `FarColorer` plugin, in some cases this leads to lots of Colorer errors in result files. E.g. found pieces of original XML file `Test.xml` are no longer valid XML in `Test.Filtered.xml` (no closing tags, etc.).

Maybe it would be better, if the result file is named like `*.txt` (neutral txt Colorer scheme) or even `{original-base-name}.{original-extension}.Filtered` hoping that this does not find any Colorer scheme. (`*.txt` is better, IMHO).

NB Technically, it's not even about `FarColorer`. Other editor tools bound to file extensions may be confused, too.

MKadaner commented:
Ouch! When I implemented this behavior, I had `.log` files in mind, but of course, all structured file formats are going to be damaged. This is totally unacceptable. Corrupted `.xml` files look awful with Colorer.

Appending `.Filtered.txt` to the original file name is a simple and safe solution. The only disadvantage, at least from my perspective, is that the original extension is lost even for non-structured files, which means coloring may change dramatically between the original and the filtered file.

Unless somebody (@HamRusTal?) suggests a more user-friendly, even if more complex to implement, solution, I'll change the behavior to `.Filtered.txt`. Will try to do it during the coming weekend.

HamRusTal commented:
> a more user-friendly, even if more complex to implement, solution

I did not look into the source code, but since the current approach involves the English word `Filtered` I assume the implementation uses FAR's l10n mechanism with `.lng` files already (and if not, it definitely should). With that in mind, I suggest that it is the entire resulting file name template that is put to the language file, with the original base name (as `{1}`) and original extension (as `{2}`) separated, like this: `{1}.{2}.Filtered.txt`. This way, those few users who are unhappy with the default will be able to customize (via `FarEng.lng.custom`) the template to their liking, e.g. `{1}_grep.{2}`.

johnd0e commented:
@MKadaner 
A couple of ideas for improvement:
1. Output Line/Column
2. Highlight found occurrences

Also, currently editor now is positioned in non-optimal way (LeftPos=CurPos), so left part of editor is hidden.

johnd0e commented:
> Or maybe a different pair of hotkeys, like Ctrl+F4 and CtrlAlt+F4

Or adding Shift to original hotkeys.
Just to clarify my intentions: the resulting file could be used for mass-editing of lines with found occurrences.
(there is a macro for that https://forum.ru-board.com/topic.cgi?forum=5&topic=49572&start=2600#19)

And for this purpose "line:" prefix would be enough, just like `grep -n`. (BTW there is another related grep options, such as `-C` for adding context lines)

> Or do you mean to highlight all found patterns? Is it possible?

Yes, I meant exactly that. It can be achieved with `ECTL_ADDCOLOR`.

alabuzhev commented:
> > * A config string with the list of extensions which should be preserved (the second format string will be used), e.g., `.txt;.log;.csv`.
> 
> Although I see your motivation, to be honest, this is not the kind of data that is supposed to reside in a language file. Maybe a new `far:config` setting instead (subject to the usual scrutiny though)? Also, do we need the leading periods? Hmm, perhaps if only to be consistent with `PATHEXT`…

To avoid new options, maybe an intermediate dialog on AltShiftF4 or something, prepopulated with the generated name, akin to ShiftF4 in panels, which the user can then invoke via a macro and change that name in any way they like?

MKadaner commented:
> Or adding Shift to original hotkeys.

`Shift` sounds good.

> Yes, I meant exactly that. It can be achieved with `ECTL_ADDCOLOR`.

I can do that. A bit lower priority. I should fix the worse things first.

> > > * A config string with the list of extensions which should be preserved (the second format string will be used), e.g., `.txt;.log;.csv`.
> > 
> > 
> > Although I see your motivation, to be honest, this is not the kind of data that is supposed to reside in a language file. Maybe a new `far:config` setting instead (subject to the usual scrutiny though)? Also, do we need the leading periods? Hmm, perhaps if only to be consistent with `PATHEXT`…
> 
> To avoid new options, maybe an intermediate dialog on AltShiftF4 or something, prepopulated with the generated name, akin to ShiftF4 in panels, which the user can then invoke via a macro and change that name in any way they like?

Yes, I meant `far:config` setting. I would really like to avoid the intermediate dialog (even if it can be handled with a macro). A dialog would break my workflow 🤣, and a macro... well, I do not like them. Is a new option really that bad?

MKadaner commented:
@johnd0e, speaking about line numbers. what do you think of adding or not adding them depending on whether they are currently visible in the menu? Kinda WYSIWYG.

Also, what do you prefer, this:

```
6: Perfect: ⒍
28: Perfect: Twenty-eight
496: Perfect: CDXCVI
```

or this:

```
  6│ Perfect: ⒍
 28│ Perfect: Twenty-eight
496│ Perfect: CDXCVI
```

HamRusTal commented:
> adding or not adding them depending on whether they are currently visible in the menu?

This is smart IMHO

> what do you prefer

What about a plain TAB character as a separator here? It is usually “long enough” to preserve alignment, is not visualized by default (so does not add noise), and is easy to remove/replace automatically.

MKadaner commented:
Thanks. I'll go with visibility then. Great!

> What about a plain TAB character as a separator here? It is usually “long enough” to preserve alignment, is not visualized by default (so does not add noise), and is easy to remove/replace automatically.

I do not agree with the “long enough” part. With Tab size set to 4 (very common setting) and 1000+ lines in the _original file_ (not uncommon), the alignment will be off:

```
1   Blah
10  Blah
100 Blah!
1000    Blah... Ouch!
```

Furthermore, I would highly prefer to right-align line numbers with spaces (just like it is done in the menu itself). It's much easier to parse visually. If the width of the line numbers column depends on the max number (the widest number fits exactly, again like in the menu), the TAB character will be adding different number of spaces (visually) depending on the size of the original file. I think this would be strange.

With line numbers right aligned, we could use space character for separator. It would still be reasonably easy to manipulate line numbers with a regex (`^\s*\d+\s`), but I strongly prefer a visible character, so that line numbers and the actual content can be told apart at a glance.

Now, the `│` character (BOX DRAWINGS LIGHT VERTICAL, U+2502) as a separator is out of question because it cannot be represented in some encodings. So, we are left with `:`

```
      1: Blah
     10: Blah
    100: Blah!
   1000: Blah... Ouch!
1000000: Blah, blah, blah, blah!
```

If there are no strong objections, I'll proceed with right-aligned, width-fit, colon separated line numbers.
